File: EaterSpring/src/main/kotlin/xyz/purema/eater/models/ProductOrder.kt
package xyz.purema.eater.models

import java.time.LocalDateTime
import java.util.*
import javax.persistence.*

@Entity
@NamedEntityGraph(
    name = "product-order-graph",
    attributeNodes = [
        NamedAttributeNode("items"),
        NamedAttributeNode("customer")
    ]
)
data class ProductOrder(
    @ManyToOne
    val customer: Customer
) {
    @Id
    val id: String = UUID.randomUUID().toString()
    val createdAt: LocalDateTime = LocalDateTime.now()

    @OneToMany(mappedBy = "productOrder", cascade = [CascadeType.ALL])
    var items: MutableSet<OrderItem> = mutableSetOf()

    fun addItem(product: Product, qty: Int) {
        items.add(
            OrderItem(
                productOrder = this,
                productId = product.id,
                name = product.name,
                price = product.price,
                qty = qty
            )
        )
    }
}

File: EaterNet/Eater/Models/ProductOrder.cs
namespace Eater.Models
{
    public class ProductOrder
    {
        public string Id { get; set; } = Guid.NewGuid().ToString();
        public DateTime CreatedAt { get; set; } = DateTime.Now;
        public Customer? Customer { get; set; }
        public ISet<OrderItem> Items { get; set; } = new HashSet<OrderItem>();

        public void AddItem(Product product, int qty)
        {
            var item = new OrderItem(
                productId: product.Id,
                name: product.Name,
                price: product.Price,
                qty: qty);

            Items.Add(item);
        }
    }
}
